[SPARK-7956] [SQL] Use Janino to compile SQL expressions into bytecode

davies opened the pull request:
In order to reduce the overhead of codegen, this PR switch to use Janino to compile SQL expressions into bytecode.

After this, the time used to compile a SQL expression is decreased from 100ms to 5ms, which is necessary to turn on codegen for general workload, also tests.

cc @rxin

Commit:
use template based string for codegen

Commit:
Merge branch 'master' of github.com:apache/spark into janino

Commit:
WIP: use Janino to compile Java source

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33698 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33698/consoleFull) for   PR 6479 at commit [`03660f3`](https://github.com/apache/spark/commit/03660f3fc14cc62647cebde76beb680d8e551afe).

SparkQA commented:
[Test build #33698 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33698/consoleFull) for   PR 6479 at commit [`03660f3`](https://github.com/apache/spark/commit/03660f3fc14cc62647cebde76beb680d8e551afe).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  case class CodeGenContext(borrowed: mutable.Stack[Expression])`
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `    class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `trait Predicate `
  * `      class SpecificPredicate implements $`
  * `abstract class BaseProject extends Projection `
  * `      class SpecificProjection extends $`
  * `      final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented on sql/catalyst/pom.xml:
```diff
@@ -60,6 +60,11 @@
       <artifactId>scalacheck_${scala.binary.version}</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
note: align this properly when you are done with hacking

rxin replied:
actually i think the proper way is to define the version in the root pom, and then here just pull janino in.

davies replied:
I'd like to have all the information of janino in one place, it's easy to manage.

JoshRosen replied:
I think that our motivation for keeping the versions in the root was to avoid having different subprojects depending on different versions of the same library.  That's probably not an issue for janino, so I guess that moving the version would only be for consistency.

rxin commented:
Any reason why you didn't use stripMargin and vertical bars? e.g.
```scala
"""|foo
   |bar
""".stripMargin
```

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala:
```diff
@@ -38,201 +44,215 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
+        }
+      }
+
+      final class SpecificRow extends ${typeOf[BaseMutableRow]} {
```
ah - i think we were discussing dropping this, didn't we?

Commit:
fix tests

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

Commit:
Merge branch 'master' of github.com:apache/spark into janino

Conflicts:
	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

davies commented:
@rxin stripMargin will remove indents in generated code, it's harder to read and debug.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33765/consoleFull) for   PR 6479 at commit [`d8f0f6c`](https://github.com/apache/spark/commit/d8f0f6c5e6f3ebd49dc7d94be83b77744ede03c7).

SparkQA commented:
[Test build #33765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33765/consoleFull) for   PR 6479 at commit [`d8f0f6c`](https://github.com/apache/spark/commit/d8f0f6c5e6f3ebd49dc7d94be83b77744ede03c7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  case class CodeGenContext(borrowed: mutable.ArrayBuffer[Expression])`
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `    class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `trait Predicate `
  * `      class SpecificPredicate implements $`
  * `abstract class BaseProject extends Projection `
  * `      class SpecificProjection extends $`
  * `      final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:
```diff
@@ -0,0 +1,178 @@
+
+import org.apache.spark.sql.types.StructType;
+
+public abstract class BaseRow implements Row {
```
Should we move this into Scala, then it will got some default implementations from Row?

davies replied:
Or, just change Row from `trait` to `abstract class`?

rxin replied:
That will break binary compatibility actually. Let me think about it a little bit more. It's unfortunate that Row wasn't an abstract class to start with.

AmplabJenkins commented:
Build finished. Test FAILed.

Commit:
fix style

Commit:
change logging level

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
remove print

SparkQA commented:
[Test build #864 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/864/consoleFull) for   PR 6479 at commit [`6089ce5`](https://github.com/apache/spark/commit/6089ce505c28868f7d0caf7717d53ebdf36649d4).

SparkQA commented:
[Test build #864 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/864/consoleFull) for   PR 6479 at commit [`6089ce5`](https://github.com/apache/spark/commit/6089ce505c28868f7d0caf7717d53ebdf36649d4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33787 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33787/consoleFull) for   PR 6479 at commit [`551a851`](https://github.com/apache/spark/commit/551a851b940616dc768ef37ba89afea15f57a295).

SparkQA commented:
[Test build #33787 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33787/consoleFull) for   PR 6479 at commit [`551a851`](https://github.com/apache/spark/commit/551a851b940616dc768ef37ba89afea15f57a295).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected case class CodeGenContext(references: mutable.ArrayBuffer[Expression])`
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `trait Predicate `
  * `      class SpecificPredicate implements $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into janino

Commit:
fix style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33929 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33929/consoleFull) for   PR 6479 at commit [`a21e968`](https://github.com/apache/spark/commit/a21e968e6b456b5c985d7c44278c4fafcb56198c).

SparkQA commented:
[Test build #33929 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33929/consoleFull) for   PR 6479 at commit [`a21e968`](https://github.com/apache/spark/commit/a21e968e6b456b5c985d7c44278c4fafcb56198c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected case class CodeGenContext(references: mutable.ArrayBuffer[Expression])`
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `trait Predicate `
  * `      class SpecificPredicate implements $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

chenghao-intel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:
```diff
@@ -0,0 +1,189 @@
+        return true;
+      }
+    }
+    float a = 0;
```
not used, should be removed.

chenghao-intel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -163,133 +190,114 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
 
       case Cast(child @ NumericType(), FloatType) =>
-        child.castOrNull(c => q"$c.toFloat", FloatType)
+        child.castOrNull(c => s"(float)($c)", FloatType)
```
This's probably not true for `Decimal` (of child), we still need to call `$c.toFloat` etc.

chenghao-intel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala:
```diff
@@ -38,201 +42,191 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
-    val code = q"""
-      final class SpecificRow(i: $rowType) extends $mutableRowType {
-        ..$classBody
+    final class SpecificRow extends ${typeOf[BaseMutableRow]} {
```
Should we give the generic row a random name and loaded by the same classloader? Or always loaded by different classloader?

davies replied:
The class is created on the fly, no classloader needed, I think.

chenghao-intel replied:
I just look at the Janino code, the classloader code probably at
https://github.com/codehaus/janino/blob/master/commons-compiler-jdk/src/org/codehaus/commons/compiler/jdk/JavaFileManagerClassLoader.java#L40

If it's has the same class name, the generated class only be loaded once, and afterward, only the first generated class(with the same class name) will be found. Sorry, I didn't run the code, let me know if I understand wrong.

davies replied:
For each compilation, cbe.cook() will create a new ByteArrayClassLoader (in SimpleCompiler.compileToClassLoader), then load the class from it. So cbe.cook() and cbe.getClazz() should guarded by a lock. It's fine to have a same name, because the name is temporary (used only inside cook()).

chenghao-intel replied:
Thanks for explanation. My concern is when those generated classes loaded in executors, will they be conflict if they have the same class name? Are they still loaded by different classloaders?
Sorry, I am not so sure how the generated class itself to be handled in spark, I will find out.

davies replied:
No, because these temporary classes are only used once.

chenghao-intel commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -36,14 +38,12 @@ class LongHashSet extends org.apache.spark.util.collection.OpenHashSet[Long]
+  protected val exprType = typeOf[Expression]
   protected val mutableRowType = typeOf[MutableRow]
   protected val genericRowType = typeOf[GenericRow]
   protected val genericMutableRowType = typeOf[GenericMutableRow]
```
Will the `SpecificMutableRow` be more optimal for primitive types?

davies replied:
`SpecificMutableRow` is generated for a given schema, the layout will be optimized (by compiler) for primitive types.

We can re-visit it later, this PR just change the code template from Scala AST to Java code.

chenghao-intel commented:
@davies any idea how the generated class to be loaded? Is it loaded / modified by Janino internally? I couldn't find any classloader relevant code in the PR.

davies commented:
@chenghao-intel The Java code is compiled by `cbe.cook(code)`, get the class `cbe.getClazz()`, then we can invoke the methods.

Commit:
Merge branch 'master' of github.com:apache/spark into janino

Commit:
fix DecimalType and cast to String

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34010 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34010/consoleFull) for   PR 6479 at commit [`f37c8c3`](https://github.com/apache/spark/commit/f37c8c37c3bab80bba62b2804cd3268a1b696377).

SparkQA commented:
[Test build #34010 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34010/consoleFull) for   PR 6479 at commit [`f37c8c3`](https://github.com/apache/spark/commit/f37c8c37c3bab80bba62b2804cd3268a1b696377).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #874 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/874/consoleFull) for   PR 6479 at commit [`f37c8c3`](https://github.com/apache/spark/commit/f37c8c37c3bab80bba62b2804cd3268a1b696377).

SparkQA commented:
[Test build #874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/874/consoleFull) for   PR 6479 at commit [`f37c8c3`](https://github.com/apache/spark/commit/f37c8c37c3bab80bba62b2804cd3268a1b696377).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected case class CodeGenContext(references: mutable.ArrayBuffer[Expression])`
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `trait Predicate `
  * `      class SpecificPredicate implements $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala:
```diff
@@ -37,7 +37,6 @@ import com.google.common.hash.Hashing
  * It uses quadratic probing with a power-of-2 hash table size, which is guaranteed
  * to explore all spaces for each key (see http://en.wikipedia.org/wiki/Quadratic_probing).
  */
-private[spark]
```
Do we have to remove this?

davies replied:
All private classes can not be accessible in Janino.

JoshRosen replied:
Ah, gotcha.  You can use the new `@Private` annotation on this class to indicate that users still shouldn't use it directly.

davies replied:
done

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -125,32 +139,45 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
+   * A context for codegen
+   * @param references the expressions that don't support codegen
+   */
+  protected case class CodeGenContext(references: mutable.ArrayBuffer[Expression])
```
I think that we can make this into a non-case class with a zero-argument constructor and a single field, `references`, that's initialized to an empty array buffer.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala:
```diff
@@ -18,18 +18,24 @@
   import scala.reflect.runtime.universe._
 
- protected def canonicalize(in: Seq[SortOrder]): Seq[SortOrder] =
+  protected def canonicalize(in: Seq[SortOrder]): Seq[SortOrder] =
```
Indentation is off here?

davies replied:
yes

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala:
```diff
@@ -18,18 +18,24 @@
-import org.apache.spark.sql.types.{BinaryType, StringType, NumericType}
+import org.apache.spark.sql.types.{BinaryType, NumericType}
+
+class BaseOrdering extends Ordering[Row] {
```
I guess this is for java-friendliness?  Can you maybe make this into an abstract class instead?

JoshRosen replied:
Can this be `private` also?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala:
```diff
@@ -20,10 +20,16 @@ package org.apache.spark.sql.catalyst.expressions.codegen
 /**
+ * Interface for generated predicate
+ */
+trait Predicate {
```
Should this be an abstract base class instead of a trait?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:
```diff
@@ -0,0 +1,187 @@
+
+  @Override
+  public boolean anyNull() {
+    for (int i=0; i< size(); i++) {
```
Nit: space after `<`.  Also, should we pull `size()` into a constant?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:
```diff
@@ -0,0 +1,187 @@
+  public StructType schema() { throw new UnsupportedOperationException(); }
+
+  @Override
+  public Object apply(int i) {
```
Since this is an alias for `get` I guess we could make it `final`.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:
```diff
@@ -0,0 +1,187 @@
+public abstract class BaseRow implements Row {
+
+  @Override
+  public int length() {
```
We can also make this `final` to ensure that it's always an alias for `size()`.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -125,32 +139,45 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
+
+  /**
+   * A context for codegen
+   * @param references the expressions that don't support codegen
```
Can you elaborate a bit more on this?  How do things get stored into this list and when do they get read back?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -163,133 +190,111 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
        * @param f a function from two primitive term names to a tree that evaluates them.
        */
-      def evaluate(f: (TermName, TermName) => Tree): Seq[Tree] =
+      def evaluate(f: (String, String) => String): String =
```
Having all of the types be `String` is a bit hard to read.  What do you think about using some type aliases for this instead?

davies replied:
Could be done later.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -313,295 +318,309 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
-          if (!${eval1.nullTerm} && ${eval1.primitiveTerm} == false) {
+        val eval1 = expressionEvaluator(e1, ctx)
+        val eval2 = expressionEvaluator(e2, ctx)
+        // TODO(davies): This is different than And.eval()
```
Can you elaborate here?  Should we file a followup JIRA?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -313,295 +318,309 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
-      case Multiply(e1, e2) => (e1, e2) evaluate { case (eval1, eval2) => q"$eval1 * $eval2" }
+        child.castOrNull(c => s"!$c", BooleanType)
+
+      case Add(e1 @ DecimalType(), e2 @ DecimalType()) =>
```
How does this match decimal types with non-default precision?  Should this be `e1: DecimalType` in order to match types of all precisions?

davies replied:
DecimalType.unapply() will catch all the DecimalType objects.

rxin replied:
@JoshRosen this is actually not correct if you use `e1: DecimalType`, because we are not matching against `DecimalType` here, but rather against expressions whose output is decimaltype.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -651,41 +677,58 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
+    case DoubleType => "double"
+    case FloatType => "float"
+    case BooleanType => "boolean"
+    case dt @ DecimalType() => "org.apache.spark.sql.types.Decimal"
```
Similar to my comments above, should this be `dt: DecimalType` in order to handle all precisions?  Or do we purposely only want to match the default precision?

davies replied:
I think it's the same, see DecimalType.unapply()

rxin replied:
It's more robust to use : DecimalType, if you just want to type check. Otherwise it could break silently when we change Decimal's unapply.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -651,41 +677,58 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
+    case BooleanType => "boolean"
+    case dt @ DecimalType() => "org.apache.spark.sql.types.Decimal"
+    case BinaryType => "byte[]"
+    case StringType => "org.apache.spark.sql.types.UTF8String"
```
We could do `classOf[UTF8String].getName` here instead.  This would save us some work if we relocate `UTF8String` to the `unsafe` package in the future.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala:
```diff
@@ -38,73 +44,89 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[Row]] wit
-    toolBox.eval(code).asInstanceOf[Ordering[Row]]
+
+    val c = compile(code)
+    val m = c.getDeclaredMethods()(0)
```
What is this code doing?

JoshRosen replied:
Might use a comment?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala:
```diff
@@ -36,41 +36,60 @@ object GenerateMutableProjection extends CodeGenerator[Seq[Expression], () => Mu
-    val code =
-      q"""
-        () => { new $mutableProjectionType {
+        $exprType[] expressions = null;
```
This could be `final`, but probably no major reason to set it final.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala:
```diff
@@ -20,10 +20,16 @@ package org.apache.spark.sql.catalyst.expressions.codegen
+ * Interface for generated predicate
+ */
+trait Predicate {
+  def predict(r: Row): Boolean
```
It might be clearer to name this `eval` instead.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala:
```diff
@@ -36,41 +36,60 @@ object GenerateMutableProjection extends CodeGenerator[Seq[Expression], () => Mu
-      q"""
-        () => { new $mutableProjectionType {
+        $exprType[] expressions = null;
+        $mutableRowType mutableRow = null;
```
This could be private.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala:
```diff
@@ -38,201 +42,191 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
-      q"if(isNullAt($iLit)) { null } else { ${newTermName(s"c$i")} }"
+      @Override
+      public Object apply(Object r) {
+        return new SpecificRow(expressions, (Row)r);
```
Space after the cast?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala:
```diff
@@ -36,41 +36,60 @@ object GenerateMutableProjection extends CodeGenerator[Seq[Expression], () => Mu
+    logDebug(s"code for ${expressions.mkString(",")}:\n$code")
+
+    val c = compile(code)
+    val m = c.getDeclaredMethods()(0)
```
Similarly, maybe add a one-line comment here to explain which method this is retrieving?

JoshRosen replied:
Is it always safe to get the declared method by position?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -651,41 +677,58 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
+    case DateType => "Integer"
+    case TimestampType => "java.sql.Timestamp"
+    case _ =>
+      "Object"
```
Nit: could go on the same line as the case.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -609,40 +628,47 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
       dataType: DataType,
       ordinal: Int,
-      value: TermName) = {
+      value: String) = {
```
Explicit return type?

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -313,295 +318,309 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
-      case MaxOf(e1, e2) =>
-        val eval1 = expressionEvaluator(e1)
-        val eval2 = expressionEvaluator(e2)
+      case MaxOf(e1, e2) if !e1.dataType.isInstanceOf[DecimalType] =>
```
Just curious: why can't this handle `DecimalType`?

davies replied:
We can not compare two Decimal by `a > b`

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -163,133 +190,111 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
-        (e1, e2).evaluateAs (BooleanType) { case (eval1, eval2) => q"$eval1 == $eval2" }
+        (e1, e2).evaluateAs (BooleanType) { case (eval1, eval2) => s"$eval1 == $eval2" }
 
       /* TODO: Fix null semantics.
```
This block of commented-out code looks like dead Scala code that can be removed in this PR.  If there's still a TODO task for the null-handling semantics, then we should file a followup JIRA.

JoshRosen commented:
There may be some additional cleanup tasks to perform in terms of deleting old code / workarounds that were needed for quasiquotes but not by Janino.  For example:

```scala
// These classes are here to avoid issues with serialization and integration with quasiquotes.
class IntegerHashSet extends org.apache.spark.util.collection.OpenHashSet[Int]
class LongHashSet extends org.apache.spark.util.collection.OpenHashSet[Long]
```

There's a `globalLock`, which I think is Scala's reflection lock, that's used in the CacheBuilder:

```scala
  protected val cache = CacheBuilder.newBuilder()
    .maximumSize(1000)
    .build(
      new CacheLoader[InType, OutType]() {
        override def load(in: InType): OutType = globalLock.synchronized {
```

Do we still need this lock?

JoshRosen commented:
More dead code to remove:

In `GenerateProjection.scala`:

```scala
    /* TODO: Configurable...
    val nullFunctions =
      s"""
        private final val nullSet = new org.apache.spark.util.collection.BitSet(length)
        final def setNullAt(i: Int) = nullSet.set(i)
        final def isNullAt(i: Int) = nullSet.get(i)
      """
     */
```

JoshRosen commented:
I took a quick pass through this PR and the changes look good at a high-level, since mostly they're just translating the quasiquotes code to its Java equivalent.  I'll loop back tomorrow and take another pass once you've addressed this round of comments.  If possible, it would be nice to see if we can completely eliminate all uses of Scala reflection and remove those imports / locks.

Commit:
address comments from Josh

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

davies commented:
@JoshRosen I think it's fine to keep the global lock, because we want to only compile once for a given expression. Also the `ClassBodyEvaluator` is not thread safe, or we need to create new one for each compile unit.

IntegerHashSet is specified version for codegen, it will have better performance than OpenHashSet[Row].

There is a follow up PR (based on this one) to push the codegen into Expression, lots of code will be moved, so some minor issues could be addressed in that one.

SparkQA commented:
[Test build #34152 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34152/consoleFull) for   PR 6479 at commit [`eec3a33`](https://github.com/apache/spark/commit/eec3a33434342bfb342fc541f97d879fbd80d974).

SparkQA commented:
[Test build #34152 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34152/consoleFull) for   PR 6479 at commit [`eec3a33`](https://github.com/apache/spark/commit/eec3a33434342bfb342fc541f97d879fbd80d974).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected class CodeGenContext `
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `abstract class Predicate `
  * `      class SpecificPredicate extends $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into janino

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34153/consoleFull) for   PR 6479 at commit [`262d848`](https://github.com/apache/spark/commit/262d84839a0876c07ffe57031fb505e664abfe66).

JoshRosen commented:
As long as we're removing quasiquotes, we might see if we can delete their dependencies from the `pom` as well:

I think we might be able to remove the following (should confirm, though):

```
 <profiles>
     <!-- Quasiquotes are merged into scala reflect from scala 2.11 onwards. -->
     <profile>
       <id>scala-2.10</id>
       <activation>
         <property><name>!scala-2.11</name></property>
       </activation>
       <dependencies>
         <dependency>
           <groupId>org.scalamacros</groupId>
           <artifactId>quasiquotes_${scala.binary.version}</artifactId>
           <version>${scala.macros.version}</version>
         </dependency>
       </dependencies>
     </profile>
   </profiles>
```

and

```
     <dependency>
       <groupId>org.scala-lang</groupId>
       <artifactId>scala-compiler</artifactId>
     </dependency>
```

We might have to keep

```
     <dependency>
       <groupId>org.scala-lang</groupId>
       <artifactId>scala-reflect</artifactId>
     </dependency>
```

if we actually depend on Scala reflection.

JoshRosen commented:
@davies, regarding the global lock, I think that we need it strictly because of `ClassBodyEvaluator`'s lack of thread-safety, not to prevent duplicate loading for expressions.

According to Guava's [CacheBuilder documentation](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/cache/CacheBuilder.html#build(com.google.common.cache.CacheLoader)) for `CacheBuilder.build(loader)`:

> Builds a cache, which either returns an already-loaded value for a given key or atomically computes or retrieves it using the supplied CacheLoader. If another thread is currently loading the value for this key, simply waits for that thread to finish and returns its loaded value. Note that multiple threads can concurrently load values for distinct keys.

Spark's HistoryServer makes use of this and I've used this pattern in some other code as well.  Not a correctness / performance issue for this PR, but just wanted to point this out since it's a neat thread-safety trick that Guava gives us.

JoshRosen commented:
Actually... if you agree that we only need the synchronization for `ClassBodyEvaluator`, how about removing `globalLock` and just synchronizing on the `ClassBodyEvaluator` instance instead?  I think that this is a little clearer / more explicit.  I think that we only need to keep the `globalLock` in place if it's implicitly guarding Scala reflection calls.

We don't need to do this cleanup here necessarily, but we might as well if it's easy to do.  The nitpicker smaller cleanups, such as type aliases for `String`, etc. should definitely be deferred, though.

rxin commented:
Does it impact correctness? If it doesn't I think it's fine to do it in followup prs.

SparkQA commented:
[Test build #34153 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34153/consoleFull) for   PR 6479 at commit [`262d848`](https://github.com/apache/spark/commit/262d84839a0876c07ffe57031fb505e664abfe66).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected class CodeGenContext `
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `abstract class Predicate `
  * `      class SpecificPredicate extends $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen commented:
Nah, not a correctness issue. Just some general cleanup; happy to tackle it myself in a followup.  By the way, Catalyst seems to compile fine when I remove both the Scala reflection and compiler JARs.  We can remove those as part of the followup, though; not a blocker for this.

Commit:
remove globalLock

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

davies commented:
@JoshRosen I had removed the globalLock, create a ClassBodyEvaluator for each compile unit (it's cheap). Removed some dependencies (introduced in codegen commit 84467468d466dadf4708a7d6a808471305149713).

SparkQA commented:
[Test build #34159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34159/consoleFull) for   PR 6479 at commit [`cc689f5`](https://github.com/apache/spark/commit/cc689f53f4e936a6c3f9721a3781971eba55f43a).

SparkQA commented:
[Test build #34159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34159/consoleFull) for   PR 6479 at commit [`cc689f5`](https://github.com/apache/spark/commit/cc689f53f4e936a6c3f9721a3781971eba55f43a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected class CodeGenContext `
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `abstract class Predicate `
  * `      class SpecificPredicate extends $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test PASSed.

davies commented:
@JoshRosen @rxin  Is this ready to go?

rxin commented:
Let's merge it.

On Thursday, June 4, 2015, Davies Liu <notifications@github.com> wrote:

> @JoshRosen <https://github.com/JoshRosen> @rxin <https://github.com/rxin>
> Is this ready to go?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/6479#issuecomment-108967511>.
>

JoshRosen commented:
LGTM.
